### PR TITLE
Adopt PR #2189: a JSON en/decoder

### PR DESCRIPTION
**Pull request by smlng**
### Contribution description

I adopted PR #2189 to get it merged, as IMHO the original contribution is still valuable to RIOT but @Kijewski (original author) hasn't found the time to work on it (no blame for that 😉). I rebased to current master and squashed reasonably, authorship was retained to stay with @Kijewski.

The original PR also included a minor adaption to embUnit, which still is a separate commit and can be dropped, if someone objects.

Otherwise this PR (was and) is in good shape and hence ready to merge.

### Issues/PRs references

- Adopts #2189

- Requires (and includes) #10538

### CODE REVIEW

File: sys/include/embUnit/AssertImpl.h [hunk 1]
```diff
@@ -46,8 +46,8 @@ void assertImplementationCStr(const char *expected,const char *actual, long line
 
 #define TEST_ASSERT_EQUAL_STRING(expected_, actual_) \
     do { \
-        __typeof__(expected_) ____expected__ = expected_; \
-        __typeof__(actual_) ____actual__ = actual_; \
+        const char *____expected__ = expected_; \
+        const char *____actual__ = actual_; \
```
Comment (miri64) [inline sys/include/embUnit/AssertImpl.h]: Why is this change required?

File: sys/include/json.h [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+ *              serialized data.
+ * @{
+ *
+ * @file
+ * @brief       Headers for the JSON module
+ *
+ * @author      René Kijewski <rene.kijewski@fu-berlin.de>
+ */
+
+#ifndef RIOT_SYS_JSON_H_
```
Comment (miri64) [inline sys/include/json.h]: Header guards are wrong according to coding conventions

File: sys/include/json.h [hunk 3]
```diff
@@ -0,0 +1,725 @@
...
+ * @see json_write_int64()
+ * @param[in] cookie The cookie that was initialized with json_write_init().
+ * @param[in] value The value to put into the data stream.
+ * @returns @ref JSON_OKAY on success.
+ * @returns @ref JSON_INVALID_DATA if the stream is inside an invalid state.
+ * @returns @ref JSON_PREMATURELY_ENDED if the data could not be written.
+ */
+json_result_t json_write_float(json_write_cookie_t *cookie, float value);
+
+#ifndef MODULE_ATMEGA_COMMON
```
Comment (miri64) [inline sys/include/json.h]: Why is this hardware dependent?

File: sys/include/json.h [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+ */
+typedef struct json_write_cookie json_write_cookie_t;
+
+/**
+ * @brief Prototype for a JSON write callback.
+ * @param[in] cookie The cookie that was originally provided to json_write_init().
+ * @param[in] buf Buffer to write.
+ * @param[in] len Length of the buffer to write.
+ * @returns `> 0` number of written bytes.
+ * @returns `< 0` to indicate an error.
```
Comment (miri64) [inline sys/include/json.h]: Here and in other places of the doc: please make this wall of text a little bit more readable (new lines between "sections", tabular alignment of command and content, ...)

File: sys/include/embUnit/AssertImpl.h (hunk 1 above)
Comment (Kijewski) [reply sys/include/embUnit/AssertImpl.h]: Something didn't work without it, but I cannot remember what.  :angel:

File: sys/include/json.h (hunk 3 above)
Comment (Kijewski) [reply sys/include/json.h]: ATMEGA did not support int64_t values at all. (In 2014, maybe my toolchain was to blame for that.)

Comment (smlng) [reply sys/include/json.h]: will check and adapt as needed

Comment (miri64) [reply sys/include/json.h]: In any case I would prefer a NOP or error message instead of excluding this function completely from the API.

**Comment by smlng** [comment]
general question: how to proceed with the requested changes? I'd like to retain the original author ship of @Kijewski, but also squashing was needed (already done here) and now need to add the fixes. Just add them as separate commits, what do you think? @miri64

**Comment by miri64** [comment]
> general question: how to proceed with the requested changes? I'd like to retain the original author ship of @Kijewski, but also squashing was needed (already done here) and now need to add the fixes. Just add them as separate commits, what do you think? @miri64

If you squash correctly, @Kijewski's authorship is retained (since ideally his commits are the base commits of your squash commits). If you want to make sure the authorship is retained, you can even use the `--author=<author>` parameter for `git commit` to set the authorship for your fixup commits. If there is a significant change (i.e. removing the platform dependency of the API) you can have it as a separate PR with yourself as an author.

**Comment by smlng** [comment]
> If you squash correctly

I'll try that 😉

**Comment by smlng** [comment]
for now I reverted my initial squashing and made some fixes to documentation and coding style. Still need to look into this atmega stuff.

**Comment by smlng** [comment]
a major problem encountered by Murdock is, that the JSON unittest increase the unittests binary size by >640K (bc it uses a JSON test suite with lots of (really large) tests strings). Thus, rendering most boards incompatible because of low FLASH/ROM size - basically we could change from blacklisting boards to whitelist `native` only. 

Any suggestions on that: @RIOT-OS/maintainers ?

**Comment by cladmi** [comment]
I would say, put it in a separate test in directly in the `tests` directory.

**Comment by cladmi** [comment]
Hmm, it may also be too big for one test alone, so maybe split it into different tests types, tests float, tests strings, tests arrays, tests dicts, and one with mixed data types.
Pthreads, threads, xtimer also have several tests folders.

**Comment by kYc0o** [comment]
@smlng what's the current size of this feature? When I reviewed the code of the former PR I found it extremely large, a lot of code for something "simple" especially the encoding. For decoding I agree it might need more lines of code.

**Comment by biboc** [comment]
What is the difference with pkg/jsmn? Is it necessary to have two json library?

**Comment by kaspar030** [comment]
> What is the difference with pkg/jsmn? Is it necessary to have two json library?

jsmn only offers very basic parsing.

**Comment by tcschmidt** [comment]
@smlng @miri64 is this a quick finish?

**Comment by smlng** [comment]
@tcschmidt code-wise: yes, otherwise not so much, because this PR adds unittests for json which are rather huge - meaning that no board can handle the complete unittests binary anymore. There is already a (yet inconclusive) discussion in #8941 (split unittests) on this matter. 

I can either split of the json unittests right away or wait for consensus in #8941.

**Comment by smlng** [comment]
hey @miri64 care to have a look at this one again?

**Comment by smlng** [comment]
Hi @miri64 thanks for looking at this again 😄 For the first: the majority of LOCs is from the test suite @Kijewski imports, it ships the generator for the test cases which is 2500++ LOCs. So splitting this into encoder/decoder doesn't help. 

Regarding the Embunit changes, I get an error when I drop the commit introducing them - it seems that the `TEST_ASSERT_EQUAL_STRING` macro expect to get a string from a macro/define too, but here they are passed in as `const char array[] = "test string"` - the modification helps. Btw. I looked into different projects that use Embunit, too and it seems that all of them have their own slight mods to make Embunit work.

**Comment by miri64** [comment]
> Regarding the Embunit changes, I get an error when I drop the commit introducing them - it seems that the `TEST_ASSERT_EQUAL_STRING` macro expect to get a string from a macro/define too, but here they are passed in as `const char array[] = "test string"` - the modification helps. Btw. I looked into different projects that use Embunit, too and it seems that all of them have their own slight mods to make Embunit work.

Ok, the let's split out at least that first.

**Comment by miri64** [comment]
Please rebase: #10538 was merged.

> Hi @miri64 thanks for looking at this again  For the first: the majority of LOCs is from the test suite @Kijewski imports, it ships the generator for the test cases which is 2500++ LOCs. So splitting this into encoder/decoder doesn't help.

I still need to review those 2500++ LOCs of test cases right ;-).

**Comment by smlng** [comment]
> Please rebase: #10538 was merged.
> 
> > Hi @miri64 thanks for looking at this again  For the first: the majority of LOCs is from the test suite @Kijewski imports, it ships the generator for the test cases which is 2500++ LOCs. So splitting this into encoder/decoder doesn't help.
> 
> I still need to review those 2500++ LOCs of test cases right ;-).

Or trust Murdock to run the test properly ...

**Comment by Kijewski** [comment]
> @miri64: I still need to review those 2500++ LOCs of test cases right ;-).

Why would you? You only need to review [/tests/json/generate_parsing_test.py](https://github.com/RIOT-OS/RIOT/blob/67946014dc5aaf695f33452aa0459933f6e771b9/tests/json/generate_parsing_test.py).

**Comment by smlng** [comment]
btw. should I try to squash this into a more reasonable number of commits already?

**Comment by miri64** [comment]
> Or trust Murdock to run the test properly ...

```C
TEST_ASSERT(true);
TEST_ASSERT(true);
TEST_ASSERT(1 == 1);
```

Success!!!1! :rofl: 

> > @miri64: I still need to review those 2500++ LOCs of test cases right ;-).
> 
> Why would you? You only need to review [/tests/json/generate_parsing_test.py](https://github.com/RIOT-OS/RIOT/blob/67946014dc5aaf695f33452aa0459933f6e771b9/tests/json/generate_parsing_test.py).

I'll try that instead ;-)

**Comment by miri64** [comment]
> btw. should I try to squash this into a more reasonable number of commits already?

Let me have a look at the current ones first (tomorrow), please.

File: tests/json/generate_parsing_test.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    return b'json_parsing_test_%s_%d' % (escaped_str, index)
+
+
+def main(dest_path='./generated-parsing-tests.c',
+         source_path='./JSONTestSuite',
+         cwd=os.path.dirname(os.path.realpath(__file__))):
+    test_names = []
+    cwd = os.path.abspath(cwd)
+    with open(os.path.join(cwd, dest_path), 'wb') as dest_file:
+        # License:
```
Comment (miri64) [inline tests/json/generate_parsing_test.py]: Every commented section can be put into its own function to make the code more readable.

File: tests/json/generate_parsing_test.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        with open(os.path.join(cwd, source_path, './LICENSE'), 'rb') as license:
+            dest_file.write(license.read())
+        dest_file.write(b'\n')
+        dest_file.write(b' */\n')
+        dest_file.write(b'\n')
+        dest_file.write(b'/* Upstream version: ???????? */\n')
+        dest_file.write(b'#error "PLEASE EDIT UPSTEAM VERSION HASH MANUALLY AND REMOVE THIS LINE!"\n')
+        dest_file.write(b'\n')
+
+        # Includes etc.:
```
Comment (miri64) [inline tests/json/generate_parsing_test.py]: Every commented section can be put into its own function to make the code more readable.

File: tests/json/generate_parsing_test.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        dest_file.write(b'#include "tests-json.h"\n')
+        dest_file.write(b'\n')
+        dest_file.write(b'#include <stddef.h>\n')
+        dest_file.write(b'\n')
+        dest_file.write(b'#pragma GCC diagnostic ignored '
+                        b'"-Woverlength-strings"\n')
+        dest_file.write(b'\n')
+
+        # Strings:
```
Comment (miri64) [inline tests/json/generate_parsing_test.py]: Every commented section can be put into its own function to make the code more readable.

File: tests/json/generate_parsing_test.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            test_name = os.path.splitext(test_name)[0]
+            test_name = test_name.encode('ascii')
+            esc_test_name = escape_test_name(test_name, index)
+
+            with open(json_name, 'rb') as input_file:
+                transform_string(dest_file, input_file, test_name,
+                                 esc_test_name)
+            test_names.append(esc_test_name)
+
+        # Strings in an array:
```
Comment (miri64) [inline tests/json/generate_parsing_test.py]: Every commented section can be put into its own function to make the code more readable.

File: tests/json/generate_parsing_test.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+         cwd=os.path.dirname(os.path.realpath(__file__))):
+    test_names = []
+    cwd = os.path.abspath(cwd)
+    with open(os.path.join(cwd, dest_path), 'wb') as dest_file:
+        # License:
+        dest_file.write(b'/* This file was generated using the tests in '
+                        b'JSON Parsing Test Suite,\n')
+        dest_file.write(b' * which has the following license information:\n')
+        dest_file.write(b'\n')
+        with open(os.path.join(cwd, source_path, './LICENSE'), 'rb') as license:
```
Comment (miri64) [inline tests/json/generate_parsing_test.py]: When starting the script without any parameters I get the following error at this line:

```
$ tests/json/generate_parsing_test.py
Traceback (most recent call last):
  File "tests/json/generate_parsing_test.py", line 124, in <module>
    result = main(*sys.argv[1:4])
  File "tests/json/generate_parsing_test.py", line 63, in main
    with open(os.path.join(cwd, source_path, './LICENSE'), 'rb') as license:
FileNotFoundError: [Errno 2] No such file or directory: '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/json/./JSONTestSuite/./LICENSE'
```

I guess this needs to be downloaded somewhere, but it would be nice to have this documented somewhere (or maybe even have a script to download it)

File: tests/json/generate_parsing_test.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        dest_file.write(b'/* Upstream version: ???????? */\n')
+        dest_file.write(b'#error "PLEASE EDIT UPSTEAM VERSION HASH MANUALLY AND REMOVE THIS LINE!"\n')
+        dest_file.write(b'\n')
+
+        # Includes etc.:
+        dest_file.write(b'#include "tests-json.h"\n')
+        dest_file.write(b'\n')
+        dest_file.write(b'#include <stddef.h>\n')
+        dest_file.write(b'\n')
+        dest_file.write(b'#pragma GCC diagnostic ignored '
```
Comment (miri64) [inline tests/json/generate_parsing_test.py]: I'm not sure, but didn't we say something against pragmas?

File: tests/json/generate_parsing_test.py (hunk 8 above)
Comment (miri64) [inline tests/json/generate_parsing_test.py]: Also: it took me a while to understand that this is basically the exposing array, not part of the text fixtures anymore. So maybe name it something different?

File: tests/json/generate_parsing_test.py (hunk 7 above)
Comment (miri64) [inline tests/json/generate_parsing_test.py]: Maybe name this function `generate_fixtures()` or something like that, to make clearer, that those strings are actually the test fixtures for the tests, not a test suite.

**Comment by miri64** [comment]
Just a question popping up in my head: any reason why we are using a self-written json library?

File: sys/include/json.h [hunk 11]
```diff
@@ -0,0 +1,859 @@
...
+ * @returns @ref JSON_OKAY on success.
+ * @returns @ref JSON_INVALID_DATA if the stream is inside an invalid state.
+ * @returns @ref JSON_PREMATURELY_ENDED if the data could not be written.
+ */
+json_result_t json_write_object_close(json_write_cookie_t *cookie);
+
+/**
+ * @brief   Write a key inside an object.
+ *
+ * Invoke this function multiple times directly to concatenate this substring.
```
Comment (miri64) [inline sys/include/json.h]: Which substring is this "this" referring to? The wording leaves it open.

File: sys/include/json.h [hunk 12]
```diff
@@ -0,0 +1,859 @@
...
+ * You could write out a string `"test"` character by character if you want to.
```
Comment (miri64) [inline sys/include/json.h]: If the first sentence would be clearer this (somewhat confusing) example wouldn't be necessary.

File: sys/include/json.h [hunk 13]
```diff
@@ -0,0 +1,859 @@
...
+ *
+ * @returns @ref JSON_OKAY on success.
+ * @returns @ref JSON_INVALID_DATA if the stream is inside an invalid state.
+ * @returns @ref JSON_PREMATURELY_ENDED if the data could not be written.
+ */
+json_result_t json_write_object_key(json_write_cookie_t *cookie, const char *string, size_t len);
+
+/**
+ * @brief   Open an array to serialize.
+ *
```
Comment (miri64) [inline sys/include/json.h]: I'm missing a detailed section similar to the one in `json_write_object open()` here.

File: sys/include/json.h [hunk 14]
```diff
@@ -0,0 +1,859 @@
...
+ * @returns @ref JSON_OKAY on success.
+ * @returns @ref JSON_INVALID_DATA if the stream is inside an invalid state.
+ * @returns @ref JSON_PREMATURELY_ENDED if the data could not be written.
+ */
+json_result_t json_write_array_open(json_write_cookie_t *cookie);
+
+/**
+ * @brief   Separate items inside the array.
+ *
+ * Inside the array call opened with json_write_array_open() call
```
Comment (miri64) [inline sys/include/json.h]: I think this belongs to `json_write_array_open()` ;-)

File: sys/include/json.h [hunk 15]
```diff
@@ -0,0 +1,859 @@
...
+ * json_write_array_next() to separate values. Calling json_write_array_next()
+ * directly after opening the array is valid but not needed.
+ * Calling json_write_array_next() before closing the array with
+ * json_write_array_close() without putting an item is invalid.
```
Comment (miri64) [inline sys/include/json.h]: From those two sentences I'm unsure if constructing empty arrays is possible.

File: sys/include/json.h [hunk 16]
```diff
@@ -0,0 +1,859 @@
...
+json_result_t json_write_array_next(json_write_cookie_t *cookie);
+
+/**
+ * @brief   Write a single integer value.
+ *
+ * JSON does not differentiate between integers and floating point numbers.
+ *
+ * @warning JSON is an unframed format. If the data to deliver consists of
+ *          a single number, then there is a fair chance that it will not be
+ *          processed correctly by the decoder side.
```
Comment (miri64) [inline sys/include/json.h]: Why is this a warning if it makes clear that something will happen *correctly*?

File: sys/include/json.h [hunk 17]
```diff
@@ -0,0 +1,859 @@
...
+ */
+json_result_t json_write_int(json_write_cookie_t *cookie, int32_t value);
+
+/**
+ * @brief   Write a floating point number.
+ *
+ * JSON does not differentiate between integers and floating point numbers.
+ *
+ * @warning The number must be finite, i.e. it must not be NaN, +Inf or -Inf.
+ *          Illegal inputs are not caught but the stream will be broken.
```
Comment (miri64) [inline sys/include/json.h]: Is there a way to write those regardless?

File: sys/include/json.h [hunk 18]
```diff
@@ -0,0 +1,859 @@
...
+ *
+ * @warning JSON is an unframed format. If the data to deliver consists of
+ *          a single number, then there is a fair chance that it will not be
+ *          processed correctly by the decoder side.
```
Comment (miri64) [inline sys/include/json.h]: Same here regarding warning about correct behavior.

File: sys/include/json.h [hunk 19]
```diff
@@ -0,0 +1,859 @@
...
+
+#ifndef MODULE_ATMEGA_COMMON
+/**
+ * @brief   Write a single integer value.
+ *
+ * JSON does not differentiate between integers and floating point numbers.
+ *
+ * @warning JSON is an unframed format. If the data to deliver consists of
+ *          a single number, then there is a fair chance that it will not be
+ *          processed correctly by the decoder side.
```
Comment (miri64) [inline sys/include/json.h]: Same here regarding warning about correct behavior.

File: sys/include/json.h [hunk 20]
```diff
@@ -0,0 +1,859 @@
...
+    JSON_READ_STATE_IN_OBJECT,   /**< The last consumed character was `{`. */
+    JSON_READ_STATE_IN_ARRAY,    /**< The last consumed character was `[`. */
+    JSON_READ_STATE_IN_SEQUENCE, /**< The stream is inside a sequence (a key, a string or a number). */
+} json_read_state_t;
+
+/**
+ * @brief   The cookie with all the information needed to deserialize JSON data.
+ *
+ * @see json_read_init()
+ * @see @ref container_of
```
Comment (miri64) [inline sys/include/json.h]: Why is this reference here?

File: sys/include/json.h (hunk 17 above)
Comment (Kijewski) [reply sys/include/json.h]: In JSON as of [RFC 7159 § 6](https://tools.ietf.org/html/rfc7159#section-6) it is not possible:

> 
   Numeric values that cannot be represented in the grammar below (such
   as Infinity and NaN) are not permitted.

File: sys/json/json_write.c [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+        default:
+            return JSON_INVALID_DATA;
+    }
+
+    cookie->state = JSON_WRITE_STATE_SOMEWHERE;
+
+    va_list ap;
+    va_start(ap, format);
+    char buffer[size];
+    int len = vsnprintf(buffer, sizeof(buffer), format, ap);
```
[the remark this answers is not in the document]
Comment (Kijewski) [reply sys/json/json_write.c]: I think the fmt module is newer than my original PR. It would make sense to use it in here.

File: sys/include/json.h (hunk 20 above)
Comment (Kijewski) [reply sys/include/json.h]: You'll most likely use this struct as a member of your own struct to keep track of it's state. If `struct json_write_cookie` is the first member, then you could simply cast the pointer, but if not you'll have to use the macro. If you think that's already obvious, then the reference may be removed.

Comment (miri64) [reply sys/include/json.h]: I think such behavior should not be encouraged ;-), so it should be removed. See also [this talk](https://media.ccc.de/v/35c3-9670-safe_and_secure_drivers_in_high-level_languages#t=184) ;-).

File: sys/json/json_write.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (smlng) [reply sys/json/json_write.c]: not right now, but will look into it soonish

File: tests/json/generate_parsing_test.py (hunk 9 above)
Comment (smlng) [reply tests/json/generate_parsing_test.py]: yep added README to provide instructions on how to update the generated stuff

**Comment by smlng** [comment]
do we (still) want this in or not? I would squash and make changes if necessary to make it work, but wouldn't "waste" more time on this for restructuring and fixing some not so nice handling. I adopted bc it looked like a nice addition.

An open point that I would address is to move the tests out of unittests bc its very big.

**Comment by miri64** [comment]
TBH, I'm not really sure. On one hand it would be nice to have a JSON encoder (though one could also do that by hand). With `jsmn` we already have a decoder however, that is less fat. On the other hand, as you said, there is definitely some fixing and rework required to get this in. I'd say: your work, your decision :-).
